src/tokens/reader_stream.rs

```diff
@@ -0,0 +1,560 @@
+// Copyright Amazon.com, Inc. or its affiliates.
+
+use super::{
+    AnnotationsThunk, ContainerType, Content, FieldNameThunk, Instruction, ScalarThunk, ScalarType,
+    ScalarValue, Token, TokenStream,
+};
+use crate::result::illegal_operation;
+use crate::thunk::Thunk;
+use crate::{IonReader, IonResult, IonType, StreamItem, Symbol};
+use std::cell::RefCell;
+use std::marker::PhantomData;
+use std::rc::Rc;
+
+const INVALID_TOKEN_ERR_TEXT: &str = "TokenStream is past this token";
+const INVALID_TOKEN_PANIC_TEXT: &str = "Invalid underlying reader state";
+
+// TODO make this more generic with respect to other readers--the problem is Item/Symbol
+// TODO this has to abstract over potentially system reader to implement macros
+
+/// Internal tracking state of the returned token.  This is necessary because the lifetime of the
+/// token returned from the stream is bound to the stream itself.
+///
+/// This is essentially strongly typed [`Option`] with convenience to operate within the
+/// context of a [`ReaderTokenStream`]
+enum UnderlyingReader<R>
+where
+    R: IonReader<Item = StreamItem, Symbol = Symbol>,
+{
+    Active(R),
+    Inactive,
+}
+
+impl<R> UnderlyingReader<R>
+where
+    R: IonReader<Item = StreamItem, Symbol = Symbol>,
+{
+    fn try_mut_ref(&mut self) -> IonResult<&mut R> {
+        use UnderlyingReader::*;
+        match self {
+            Active(reader) => Ok(reader),
+            Inactive => illegal_operation(INVALID_TOKEN_ERR_TEXT),
+        }
+    }
+
+    fn try_ref(&self) -> IonResult<&R> {
+        use UnderlyingReader::*;
+        match self {
+            Active(reader) => Ok(reader),
+            Inactive => illegal_operation(INVALID_TOKEN_ERR_TEXT),
+        }
+    }
+
+    fn as_mut_ref(&mut self) -> &mut R {
+        self.try_mut_ref().expect(INVALID_TOKEN_PANIC_TEXT)
+    }
+
+    fn as_ref(&self) -> &R {
+        self.try_ref().expect(INVALID_TOKEN_PANIC_TEXT)
+    }
+
+    /// Invalidates this reader returning its original underlying reader.
+    fn invalidate(&mut self) -> Self {
+        std::mem::replace(self, UnderlyingReader::Inactive)
+    }
+}
+
+/// Indicates the state of the last token returned.
+#[derive(Debug, Eq, PartialEq, Copy, Clone)]
+enum PendingToken {
+    Materialized,
+    Deferred,
+    None,
+}
+
+impl PendingToken {
+    fn is_deferred(&self) -> bool {
+        matches!(self, PendingToken::Deferred)
+    }
+}
+
+/// Condensed version of the item of the [`IonReader`].
+#[derive(Debug, Eq, PartialEq, Copy, Clone)]
+enum ReaderPosition {
+    /// We are positioned on nothing.
+    Nothing,
+    /// We are positioned on some value.
+    NonContainer,
```

Comment:
Is this "non-container" (vs "scalar") because it includes null containers, which are arguably scalar values even though their Ion type is a container? If so, that'd be helpful to have in the comments.